lib/modules/ai_bot/personas/persona.rb

```diff
@@ -0,0 +1,119 @@
+          I18n.t("discourse_ai.ai_bot.personas.#{to_s.demodulize.underscore}.description")
+        end
+
+        def commands
+          []
+        end
+
+        def render_commands(render_function_instructions:)
+          result = +""
+          if render_function_instructions
+            result << "\n"
+            result << function_list.system_prompt
+            result << "\n"
+          end
+          result << available_commands.map(&:custom_system_message).compact.join("\n")
+          result
+        end
+
+        def render_system_prompt(topic: nil, render_function_instructions: true)
+          substitutions = {
+            site_url: Discourse.base_url,
+            site_title: SiteSetting.title,
+            site_description: SiteSetting.site_description,
+            time: Time.zone.now,
+            commands: render_commands(render_function_instructions: render_function_instructions),
+          }
+
+          substitutions[:participants] = topic.allowed_users.map(&:username).join(", ") if topic
+
+          system_prompt.gsub(/\{(\w+)\}/) do |match|
+            found = substitutions[match[1..-2].to_sym]
+            found.nil? ? match : found.to_s
+          end
+        end
+
+        def available_commands
+          return @available_commands if @available_commands
+
+          @available_commands = all_available_commands.filter { |cmd| commands.include?(cmd) }
+        end
+
+        def available_functions
+          # note if defined? can be a problem in test
+          # this can never be nil so it is safe
+          return @available_functions if @available_functions
+
+          functions = []
+
+          functions =
+            available_commands.map do |command|
+              function =
+                DiscourseAi::Inference::Function.new(name: command.name, description: command.desc)
+              command.parameters.each do |parameter|
+                function.add_parameter(
+                  name: parameter.name,
+                  type: parameter.type,
+                  description: parameter.description,
+                  required: parameter.required,
+                  enum: parameter.enum,
+                )
+              end
+              function
+            end
+
+          @available_functions = functions
+        end
+
+        def function_list
+          return @function_list if @function_list
+
+          @function_list = DiscourseAi::Inference::FunctionList.new
+          available_functions.each { |function| @function_list << function }
+          @function_list
+        end
+
+        def all_available_commands
+          return @cmds if @cmds
+
+          all_commands =
+            [
+              Commands::CategoriesCommand,
+              Commands::TimeCommand,
+              Commands::SearchCommand,
+              Commands::SummarizeCommand,
+              Commands::ReadCommand,
+              Commands::SettingContextCommand,
+            ].tap do |cmds|
```

Comment:
Personal preference but I find the use of `tap` with the array here makes it a bit harder to read (whereas I wouldn't think twice about it for class construction(

Reply:
yeah not a fan of tap, this did not come from me :) will change